crates/pipeline-manager/src/compiler/rust_compiler.rs

```diff
@@ -999,82 +1002,143 @@ async fn call_compiler(
+impl From<UtilError> for RustCompilationCleanupError {
+    fn from(value: UtilError) -> Self {
+        RustCompilationCleanupError::Utility(value)
     }
 }
 
 /// Makes the cleanup decision based on the provided name of the file or directory.
 fn decide_cleanup(
     name: &str,
-    project_separator: char,
     remainder_separator: Option<(bool, char)>,
     deletion: &[String],
 ) -> CleanupDecision {
-    // Name must start with "project<project_separator>" (e.g. "project-", "project_")
-    // Remainder is for example: "project-0123456789.abcdef" -> "0123456789.abcdef"
-    let Some(remainder) = name.strip_prefix(&format!("project{project_separator}")) else {
+    // Ignore anything which does not start with "feldera_pipe_" or "libfeldera_pipe_"
+    if !name.starts_with("feldera_pipe_") && !name.starts_with("libfeldera_pipe_") {
         return CleanupDecision::Ignore;
-    };
+    }
 
     // The checksum needs to be retrieved from the remainder
-    let checksum = match remainder_separator {
-        None => remainder,
+    let core_name = match remainder_separator {
+        None => name,
         // The remainder split can either be required or optional.
         // If it is required, it will not return the entire remainder as default
         // if it cannot split the remainder by the separator.
         Some((required, separator)) => {
-            let remainder_spl: Vec<&str> = remainder.splitn(2, separator).collect();
+            let remainder_spl: Vec<&str> = name.splitn(2, separator).collect();
             if remainder_spl.len() == 2 {
                 remainder_spl[0]
             } else if required {
+                // If remainder is required for it to be considered,
+                // it will be ignored as it doesn't match the criteria
                 return CleanupDecision::Ignore;
             } else {
-                remainder
+                name
             }
         }
     };
 
     // Final decision whether to keep or not is determined whether it is marked for deletion
-    if deletion.contains(&checksum.to_string()) {
+    if deletion.contains(&core_name.to_string()) {
         CleanupDecision::Remove
     } else {
         CleanupDecision::Keep {
-            motivation: checksum.to_string(),
+            motivation: core_name.to_string(),
         }
     }
 }
 
-/// Cleans up the Rust compilation working directory by removing projects
-/// and binaries of pipelines that no longer exist. It also removes files
-/// and directories that are unexpected.
+/// Cleans up the Rust compilation working directory by removing binaries
+/// and compilation artifacts of pipeline programs that no longer exist.
 async fn cleanup_rust_compilation(
     config: &CompilerConfig,
     db: Arc<Mutex<StoragePostgres>>,
 ) -> Result<(), RustCompilationCleanupError> {
     trace!("Performing Rust cleanup...");
 
+    // Rust compilation directory
+    let rust_compilation_dir = config.working_dir().join("rust-compilation");
```

Comment:
can races trigger this function to run concurrently with itself?

Reply:
No, only one compiler server is alive at a time, and it strictly interleaves compilation with cleanup. As such, either compilation is running, or cleanup, or nothing.